libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs

```diff
@@ -85,88 +97,121 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
 
             // Format list of choices
-            var connector = string.Empty;
-            var txt = text ?? string.Empty;
-            txt += " ";
-
-            for (var index = 0; index < choices.Count; index++)
+            var separator = string.Empty;
+            var sb = new StringBuilder();
+            if (text != null)
             {
-                var choice = choices[index];
-
-                var title = choice.Action != null && choice.Action.Title != null ? choice.Action.Title : choice.Value;
+                sb.Append(text);
+            }
+            sb.Append(' ');
 
-                txt += $"{connector}";
-                if (opt.IncludeNumbers.Value)
-                {
-                    txt += "(" + (index + 1).ToString() + ") ";
-                }
-                txt += $"{title}";
-                if (index == (choices.Count - 2))
+            var list = choices.ToList();
+            for (int i = 0; i < list.Count; i++)
+            {
+                var choice = list[i];
+                if (choice is null)
                 {
-                    connector = (index == 0 ? opt.InlineOr : opt.InlineOrMore) ?? string.Empty;
+                    throw new ArgumentException("Set of choices contained a null element, which is not allowed.", nameof(choices));
                 }
-                else
+
+                string title = GetTitle(choice);
+                sb.Append(separator);
+                if (opt.IncludeNumbers.Value)
                 {
-                    connector = opt.InlineSeparator ?? string.Empty;
+                    sb.Append('(');
+                    sb.Append(i + 1);
+                    sb.Append(") ");
                 }
+                sb.Append(title);
+
+                // Insert separators between items.
+                // Use an "Oxford comma"-style separator for three or more items.
+                separator = (i == list.Count - 2)
+                    ? (i == 0 ? opt.InlineOr : opt.InlineOrMore) ?? string.Empty
+                    : opt.InlineSeparator ?? string.Empty;
             }
-            txt += "";
 
             // Return activity with choices as an inline list.
-            return MessageFactory.Text(txt, speak, InputHints.ExpectingInput);
+            return MessageFactory.Text(sb.ToString(), speak, InputHints.ExpectingInput);
         }
 
-        public static Activity List(List<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        /// <summary>
+        /// Gets a "normalized" title for a <see cref="Choice"/> object.
+        /// </summary>
+        /// <param name="choice">The choice object.</param>
+        /// <returns>The normalized title.</returns>
+        /// <exception cref="ArgumentNullException"/>
+        /// <exception cref="InvalidOperationException">The choice does not have a valid action title or value to use as a title.</exception>
+        private static string GetTitle(Choice choice)
```

Comment:
I like the factoring of code here, the way this was extracted. 

... but let's try to keep each change seperate and tested.